src/MongoDB.Driver.Core/Core/WireProtocol/Messages/Encoders/JsonEncoders/CompressedMessageJsonEncoder.cs

```diff
@@ -0,0 +1,166 @@
+
+                return new CompressedMessage(originalMessage, null, compressorId);
+            }
+        }
+
+        /// <summary>
+        /// Writes the message.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        public void WriteMessage(CompressedMessage message)
+        {
+            Ensure.IsNotNull(message, nameof(message));
+
+            var writer = CreateJsonWriter();
+
+            writer.WriteStartDocument();
+            writer.WriteInt32("opcode", (int)Opcode.Compressed);
+            writer.WriteInt32("compressorId", (int)message.CompressorType);
+            writer.WriteInt32("originalOpcode", (int)GetOpcodeByDbMessageType(message.OriginalMessage.MessageType));
+            using (var originalWriter = new StringWriter())
+            {
+                var encoder = GetEncoderByOpcode(Opcode.Update, textWriter: originalWriter);
+                encoder.WriteMessage(message.OriginalMessage);
+                writer.WriteString("compressedMessage", originalWriter.ToString());
+            }
+
+            writer.WriteEndDocument();
+        }
+
+        /// <inheritdoc />
+        MongoDBMessage IMessageEncoder.ReadMessage()
+        {
+            return ReadMessage();
+        }
+
+        /// <inheritdoc />
+        void IMessageEncoder.WriteMessage(MongoDBMessage message)
+        {
+            WriteMessage((CompressedMessage)message);
+        }
+
+        private IMessageEncoder GetEncoderByOpcode(Opcode opcode, TextReader textReader = null, TextWriter textWriter = null)
+        {
+            var encoderFactory = new JsonMessageEncoderFactory(textReader, textWriter, _encoderSettings);
+            switch (opcode)
+            {
+                // NOTE: Opcode.Message is never used
+                case Opcode.Reply:
+                    return encoderFactory.GetReplyMessageEncoder(RawBsonDocumentSerializer.Instance);
+                case Opcode.OpMsg:
+                    return encoderFactory.GetCommandResponseMessageEncoder();
+                case Opcode.Delete:
+                    return encoderFactory.GetDeleteMessageEncoder();
+                case Opcode.GetMore:
+                    return encoderFactory.GetGetMoreMessageEncoder();
+                case Opcode.Insert:
+                    return encoderFactory.GetInsertMessageEncoder(RawBsonDocumentSerializer.Instance);
+                case Opcode.Update:
+                    return encoderFactory.GetUpdateMessageEncoder();
+                case Opcode.Query:
+                    return encoderFactory.GetQueryMessageEncoder();
+                case Opcode.KillCursors:
+                    return encoderFactory.GetKillCursorsMessageEncoder();
+            }
+
+            throw new ArgumentException($"The opcode '{opcode}' is not allowed.");
+        }
+
+        private Opcode GetOpcodeByDbMessageType(MongoDBMessageType mongoDbMessageType)
+        {
+            switch (mongoDbMessageType)
```

Comment:
Maybe we can leave only one enum either `MongoDbMessageType` or `Opcode` on the driver side?